File: output/GPT3.5-Turbo/xCodeEval_remove_4000/Java/Python/a2105f3a6efbd03893ab23ee4b039ac3.py
import math
import sys

class F3:
    def __init__(self):
        self.INPUT = ""
        self.isp = []
        self.xcums = []
        self.primes = []
        self.magic = []
        self.id3 = []
        self.id7 = None
        self.picache = None
        self.id1 = None
        self.cumps = []
        self.FP = []
        self.M = 0
        self.B = 0
        self.cachelimit = 0

    def id0(self, n):
        if n <= 32:
            primes = [2, 3, 5, 7, 11, 13, 17, 19, 23, 29, 31]
            for i in range(len(primes)):
                if n < primes[i]:
                    return primes[:i]
            return primes

        u = n + 32
        lu = math.log(u)
        ret = [0] * int(u / lu + u / lu / lu * 1.5)
        ret[0] = 2
        pos = 1

        isnp = [0] * ((n + 1) // 32 // 2 + 1)
        sup = (n + 1) // 32 // 2 + 1

        for tp in self.id3:
            ret[pos] = tp
            pos += 1
            ptn = [0] * tp
            for i in range((tp - 3) // 2, tp << 5, tp):
                ptn[i >> 5] |= 1 << (i & 31)
            for j in range(0, sup, tp):
                for i in range(tp):
                    if i + j < sup:
                        isnp[j + i] |= ptn[i]

        h = n // 2
        for i in range(sup):
            for j in range(~isnp[i] & -isnp[i]):
                pp = i << 5 | self.magic[(j * 0x076be629) >> 27]
                p = 2 * pp + 3
                if p > n:
                    break
                ret[pos] = p
                pos += 1
                if p * p > n:
                    continue
                for q in range((p * p - 3) // 2, h + 1, p):
                    isnp[q >> 5] |= 1 << q

        return ret[:pos]

    def solve(self):
        n = self.nl()
        ans = 0
        if n <= 10:
            if n >= 6:
                ans += 1
            if n >= 8:
                ans += 1
            if n >= 10:
                ans += 1
            print(ans)
            return

        ml = self.id7(n // 2)
        for i in range(len(self.primes)):
            plus = ml.pi(n // self.primes[i]) - (i + 1)
            if plus <= 0:
                break
            ans += plus

        for p in self.primes:
            if p * p * p <= n:
                ans += 1
            else:
                break
        print(ans)

    class id7:
        def __init__(self, n):
            self.primes = []
            self.xcums = []
            self.isp = []
            self.cachelimit = 0
            self.picache = None
            self.id1 = None
            self.cachePhi = False
            self.cumps = []
            self.FP = []
            self.M = 0
            self.B = 0

            assert n >= 4

            self.picache = self.id4()
            self.id1 = self.id4()

            s = int(math.sqrt(n)) * self.B

            self.cachelimit = s
            self.primes = self.id0(s)
            self.xcums = [0] * ((s >> 6) + 1)
            self.isp = [0] * ((s >> 6) + 1)
            for p in self.primes:
                self.isp[p >> 6] |= 1 << p
            for i in range(len(self.xcums) - 1):
                self.xcums[i + 1] = self.xcums[i] + bin(self.isp[i]).count('1')

            self.cumps = []
            b = [True]
            length = 1
            for i in range(self.M):
                length *= self.FP[i]
                c = [False] * length
                for j in range(self.FP[i]):
                    c[j * len(b):(j + 1) * len(b)] = b
                for k in range(0, length, self.FP[i]):
                    c[k] = False
                cumps = [0] * (length + 1)
                for j in range(length):
                    cumps[j + 1] = cumps[j] + (1 if c[j] else 0)
                self.FP[i] = length
                b = c
                self.cumps.append(cumps)

        def sqrt(self, n):
            i = max(0, int(math.sqrt(n)) - 2)
            while i * i <= n:
                i += 1
            return i - 1

        def id6(self, n):
            i = max(0, int(math.id6(n)) - 2)
            while i * i * i <= n:
                i += 1
            return i - 1

        def pi(self, x):
            if x <= self.cachelimit:
                ix = int(x)
                return self.xcums[ix >> 6] + bin(self.isp[ix >> 6] << ~ix).count('1')

            if self.picache.containsKey(x):
                return self.picache.get(x)

            A = self.pi(self.sqrt(self.sqrt(x)))
            ret = A + self.phi(x, A) - self.P2(x, A) - self.P3(x, A) - 1
            self.picache.put(x, ret)
            return ret

        def phi(self, x, A):
            if 0 < A - 1 < self.M:
                return self.cumps[A - 1][self.FP[A - 1]] * (x // self.FP[A - 1]) + self.cumps[A - 1][int(x % self.FP[A - 1] + 1)]
            if A > 0 and x <= self.primes[A - 1] * self.primes[A - 1]:
                return self.pi(x) - A + 1

            ret = x
            for i in range(A - 1, -1, -1):
                ret -= self.phi(x // self.primes[i], i)
            return ret

        def P2(self, x, A):
            B = self.pi(self.sqrt(x))
            ret = 0
            for i in range(A, B):
                ret += self.pi(x // self.primes[i])
            ret -= (B - A) * (B + A - 1) // 2
            return ret

        def P3(self, x, A):
            C = self.pi(self.id6(x))
            ret = 0
            for i in range(A, C):
                xi = x // self.primes[i]
                B = self.pi(self.sqrt(xi))
                for j in range(i, B):
                    ret += self.pi(xi // self.primes[j]) - j
            return ret

        class id4:
            def __init__(self):
                self.keys = []
                self.allocated = []
                self.scale = 1 << 2
                self.rscale = 1 << 1
                self.mask = self.scale - 1
                self.size = 0

                self.allocated = [0] * self.scale
                self.keys = [0] * self.scale

            def containsKey(self, x):
                pos = self.h(x) & self.mask
                while self.allocated[pos] != 0:
                    if x == self.keys[pos]:
                        return True
                    pos = (pos + 1) & self.mask
                return False

            def get(self, x):
                pos = self.h(x) & self.mask
                while self.allocated[pos] != 0:
                    if x == self.keys[pos]:
                        return self.allocated[pos]
                    pos = (pos + 1) & self.mask
                return 0

            def put(self, x, v):
                pos = self.h(x) & self.mask
                while self.allocated[pos] != 0:
                    if x == self.keys[pos]:
                        oldval = self.allocated[pos]
                        self.allocated[pos] = v
                        return oldval
                    pos = (pos + 1) & self.mask
                if self.size == self.rscale:
                    self.id5(x, v)
                else:
                    self.keys[pos] = x
                    self.allocated[pos] = v
                self.size += 1
                return 0

            def inc(self, x, v):
                pos = self.h(x) & self.mask
                while self.allocated[pos] != 0:
                    if x == self.keys[pos]:
                        self.allocated[pos] += v
                        return self.allocated[pos]
                    pos = (pos + 1) & self.mask
                if self.size == self.rscale:
                    self.id5(x, v)
                else:
                    self.keys[pos] = x
                    self.allocated[pos] = v
                self.size += 1
                return v

            def remove(self, x):
                pos = self.h(x) & self.mask
                while self.allocated[pos] != 0:
                    if x == self.keys[pos]:
                        self.size -= 1

                        last = pos
                        pos = (pos + 1) & self.mask
                        while self.allocated[pos] != 0:
                            lh = self.h(self.keys[pos]) & self.mask

                            if lh <= last < pos or pos < lh <= last or last < pos < lh:
                                self.keys[last] = self.keys[pos]
                                self.allocated[last] = self.allocated[pos]
                                last = pos
                            pos = (pos + 1) & self.mask
                        self.keys[last] = 0
                        self.allocated[last] = 0

                        return True
                    pos = (pos + 1) & self.mask
                return False

            def id5(self, x, v):
                nscale = self.scale << 1
                nrscale = self.rscale << 1
                nmask = nscale - 1
                nallocated = [0] * nscale
                nkeys = [0] * nscale
                for i in self.next(0):
                    y = self.keys[i]
                    pos = self.h(y) & nmask
                    while nallocated[pos] != 0:
                        pos = (pos + 1) & nmask
                    nkeys[pos] = y
                    nallocated[pos] = self.allocated[i]

                pos = self.h(x) & nmask
                while nallocated[pos] != 0:
                    pos = (pos + 1) & nmask
                nkeys[pos] = x
                nallocated[pos] = v

                self.allocated = nallocated
                self.keys = nkeys
                self.scale = nscale
                self.rscale = nrscale
                self.mask = nmask

            def next(self, itr):
                while itr < self.scale and self.allocated[itr] == 0:
                    itr += 1
                return itr

            def h(self, x):
                x ^= x >> 33
                x *= 0xff51afd7ed558ccd
                x ^= x >> 33
                x *= 0xc4ceb9fe1a85ec53
                x ^= x >> 33
                return int(x)

            def __str__(self):
                sb = []
                for i in self.next(0):
                    sb.append(",")
                    sb.append(str(self.keys[i]) + ":" + str(self.allocated[i]))
                return "" if len(sb) == 0 else sb[1:]

    def run(self):
        self.is = sys.stdin
        self.out = sys.stdout

        self.solve()
        self.out.flush()

    def main(self):
        self.run()

    def readByte(self):
        c = sys.stdin.read(1)
        return -1 if not c else ord(c)

    def id2(self, c):
        return not (33 <= c <= 126)

    def skip(self):
        b = self.readByte()
        while b != -1 and not self.id2(b):
            b = self.readByte()
        return b

    def nd(self):
        return float(self.ns())

    def nc(self):
        return chr(self.skip())

    def ns(self):
        b = self.skip()
        sb = []
        while not self.id2(b):
            sb.append(chr(b))
            b = self.readByte()
        return "".join(sb)

    def ns(self, n):
        buf = [0] * n
        b = self.skip()
        p = 0
        while p < n and not self.id2(b):
            buf[p] = chr(b)
            p += 1
            b = self.readByte()
        return buf if n == p else buf[:p]

    def nm(self, n, m):
        return [self.ns(m) for _ in range(n)]

    def na(self, n):
        return [int(x) for x in self.ns().split()]

    def ni(self):
        num = 0
        b = self.readByte()
        minus = False
        while b != -1 and not (48 <= b <= 57 or b == 45):
            b = self.readByte()
        if b == 45:
            minus = True
            b = self.readByte()

        while 48 <= b <= 57:
            num = num * 10 + (b - 48)
            b = self.readByte()
        return -num if minus else num

    def nl(self):
        num = 0
        b = self.readByte()
        minus = False
        while b != -1 and not (48 <= b <= 57 or b == 45):
            b = self.readByte()
        if b == 45:
            minus = True
            b = self.readByte()

        while 48 <= b <= 57:
            num = num * 10 + (b - 48)
            b = self.readByte()
        return -num if minus else num

    def tr(self, *o):
        if not self.oj:
            print(*o)

if __name__ == "__main__":
    F3().main()

File: output/GPT3.5-Turbo/xCodeEval_remove_4000/Java/Go/a2105f3a6efbd03893ab23ee4b039ac3.go
package main

import (
	"bufio"
	"fmt"
	"math"
	"os"
	"strconv"
	"strings"
)

var (
	scanner *bufio.Scanner
	writer  *bufio.Writer
)

func initIO() {
	scanner = bufio.NewScanner(os.Stdin)
	scanner.Split(bufio.ScanWords)
	writer = bufio.NewWriter(os.Stdout)
}

func flush() {
	writer.Flush()
}

func readString() string {
	scanner.Scan()
	return scanner.Text()
}

func readInt() int {
	i, _ := strconv.Atoi(readString())
	return i
}

func readInt64() int64 {
	i, _ := strconv.ParseInt(readString(), 10, 64)
	return i
}

func readFloat64() float64 {
	i, _ := strconv.ParseFloat(readString(), 64)
	return i
}

func writeString(s string) {
	writer.WriteString(s)
}

func writeInt(i int) {
	writeString(strconv.Itoa(i))
}

func writeInt64(i int64) {
	writeString(strconv.FormatInt(i, 10))
}

func writeFloat64(f float64) {
	writeString(strconv.FormatFloat(f, 'f', -1, 64))
}

func solve() {
	n := readInt64()
	ans := int64(0)
	if n <= 10 {
		if n >= 6 {
			ans++
		}
		if n >= 8 {
			ans++
		}
		if n >= 10 {
			ans++
		}
		writeInt(int(ans))
		return
	}

	primes := id0(400000)
	ml := id7(n / 2)
	for i := 0; i < len(primes); i++ {
		plus := ml.pi(n/primes[i]) - int64(i+1)
		if plus <= 0 {
			break
		}
		ans += plus
	}

	for _, p := range primes {
		if int64(p)*int64(p)*int64(p) <= n {
			ans++
		} else {
			break
		}
	}
	writeInt(int(ans))
}

func main() {
	initIO()
	defer flush()
	solve()
}

func id0(n int) []int {
	if n <= 32 {
		primes := []int{2, 3, 5, 7, 11, 13, 17, 19, 23, 29, 31}
		for i := 0; i < len(primes); i++ {
			if n < primes[i] {
				return primes[:i]
			}
		}
		return primes
	}

	u := n + 32
	lu := math.Log(float64(u))
	ret := make([]int, int(u/lu+u/lu/lu*1.5))
	ret[0] = 2
	pos := 1

	isnp := make([]int, (n+1)/32/2+1)
	sup := (n + 1) / 32 / 2 + 1

	id3 := []int{3, 5, 7, 11, 13, 17, 19, 23, 29, 31}
	for _, tp := range id3 {
		ret[pos] = tp
		pos++
		ptn := make([]int, tp)
		for i := (tp - 3) / 2; i < tp<<5; i += tp {
			ptn[i>>5] |= 1 << (i & 31)
		}
		for j := 0; j < sup; j += tp {
			for i := 0; i < tp && i+j < sup; i++ {
				isnp[j+i] |= ptn[i]
			}
		}
	}

	magic := []int{0, 1, 23, 2, 29, 24, 19, 3, 30, 27, 25, 11, 20, 8, 4, 13, 31, 22, 28, 18, 26, 10, 7, 12, 21, 17, 9, 6, 16, 5, 15, 14}
	h := n / 2
	for i := 0; i < sup; i++ {
		for j := ^isnp[i]; j != 0; j &= j - 1 {
			pp := i<<5 | magic[(j&-j)*0x076be629>>27]
			p := 2*pp + 3
			if p > n {
				break
			}
			ret[pos] = p
			pos++
			if int64(p)*int64(p) > n {
				continue
			}
			for q := (p*p - 3) / 2; q <= h; q += p {
				isnp[q>>5] |= 1 << q
			}
		}
	}

	return ret[:pos]
}

type id7 struct {
	primes []int
	xcums  []int
	isp    []int64
	cumps  [][]int
	FP     []int
	M      int
	B      int
}

func newID7(n int64) *id7 {
	id7 := &id7{}
	id7.primes = id0(int(n))
	id7.xcums = make([]int, (len(id7.primes)>>6)+1)
	id7.isp = make([]int64, (len(id7.primes)>>6)+1)
	for _, p := range id7.primes {
		id7.isp[p>>6] |= 1 << p
	}
	for i := 0; i < len(id7.xcums)-1; i++ {
		id7.xcums[i+1] = id7.xcums[i] + int64BitsCount(id7.isp[i])
	}

	id7.cumps = make([][]int, id7.M)
	b := []bool{true}
	len := 1
	for i := 0; i < id7.M; i++ {
		len *= id7.FP[i]
		c := make([]bool, len)
		for j := 0; j < id7.FP[i]; j++ {
			copy(c[j*len:], b)
		}
		for k := 0; k < len; k += id7.FP[i] {
			c[k] = false
		}
		id7.cumps[i] = make([]int, len+1)
		for j := 0; j < len; j++ {
			if c[j] {
				id7.cumps[i][j+1] = id7.cumps[i][j] + 1
			} else {
				id7.cumps[i][j+1] = id7.cumps[i][j]
			}
		}
		id7.FP[i] = len
		b = c
	}

	return id7
}

func (id7 *id7) pi(x int64) int64 {
	if x <= int64(id7.xcums[len(id7.xcums)-1]) {
		ix := int(x)
		return int64(id7.xcums[ix>>6]) + int64BitsCount(id7.isp[ix>>6]<<^ix)
	}
	if picache.containsKey(x) {
		return picache.get(x)
	}

	A := int(id7.pi(sqrt(sqrt(x))))
	ret := int64(A + phi(x, A) - P2(x, A) - P3(x, A) - 1)
	picache.put(x, ret)
	return ret
}

func phi(x int64, A int) int64 {
	if A > 0 && A-1 < id7.M {
		return int64(id7.cumps[A-1][id7.FP[A-1]])*(x/int64(id7.FP[A-1])) + int64(id7.cumps[A-1][int(x%int64(id7.FP[A-1])+1)])
	}
	if A > 0 && x <= int64(id7.primes[A-1])*int64(id7.primes[A-1]) {
		return id7.pi(x) - int64(A) + 1
	}

	ret := x
	for i := A - 1; i >= 0; i-- {
		ret -= phi(x/int64(id7.primes[i]), i)
	}

	return ret
}

func P2(x int64, A int) int64 {
	B := int(id7.pi(sqrt(x)))
	ret := int64(0)
	for i := A; i < B; i++ {
		ret += id7.pi(x / int64(id7.primes[i]))
	}
	ret -= int64(B-A)*(int64(B)+int64(A)-1)/2
	return ret
}

func P3(x int64, A int) int64 {
	C := int(id7.pi(id6(x)))
	ret := int64(0)
	for i := A; i < C; i++ {
		xi := x / int64(id7.primes[i])
		B := int(id7.pi(sqrt(xi)))
		for j := i; j < B; j++ {
			ret += id7.pi(xi / int64(id7.primes[j])) - int64(j)
		}
	}
	return ret
}

func int64BitsCount(x int64) int {
	count := 0
	for x != 0 {
		x &= x - 1
		count++
	}
	return count
}

type id4 struct {
	keys      []int64
	allocated []int64
	scale     int
	rscale    int
	mask      int
	size      int
}

func newID4() *id4 {
	id4 := &id4{}
	id4.scale = 1 << 2
	id4.rscale = 1 << 1
	id4.mask = id4.scale - 1
	id4.allocated = make([]int64, id4.scale)
	for i := 0; i < id4.scale; i++ {
		id4.allocated[i] = NG
	}
	id4.keys = make([]int64, id4.scale)
	return id4
}

func (id4 *id4) containsKey(x int64) bool {
	pos := h(x) & id4.mask
	for id4.allocated[pos] != NG {
		if x == id4.keys[pos] {
			return true
		}
		pos = (pos + 1) & id4.mask
	}
	return false
}

func (id4 *id4) get(x int64) int64 {
	pos := h(x) & id4.mask
	for id4.allocated[pos] != NG {
		if x == id4.keys[pos] {
			return id4.allocated[pos]
		}
		pos = (pos + 1) & id4.mask
	}
	return NG
}

func (id4 *id4) put(x, v int64) int64 {
	pos := h(x) & id4.mask
	for id4.allocated[pos] != NG {
		if x == id4.keys[pos] {
			oldval := id4.allocated[pos]
			id4.allocated[pos] = v
			return oldval
		}
		pos = (pos + 1) & id4.mask
	}
	if id4.size == id4.rscale {
		id4.id5(x, v)
	} else {
		id4.keys[pos] = x
		id4.allocated[pos] = v
	}
	id4.size++
	return 0
}

func (id4 *id4) inc(x, v int64) int64 {
	pos := h(x) & id4.mask
	for id4.allocated[pos] != NG {
		if x == id4.keys[pos] {
			id4.allocated[pos] += v
			return id4.allocated[pos]
		}
		pos = (pos + 1) & id4.mask
	}
	if id4.size == id4.rscale {
		id4.id5(x, v)
	} else {
		id4.keys[pos] = x
		id4.allocated[pos] = v
	}
	id4.size++
	return v
}

func (id4 *id4) remove(x int64) bool {
	pos := h(x) & id4.mask
	for id4.allocated[pos] != NG {
		if x == id4.keys[pos] {
			id4.size--

			last := pos
			pos = (pos + 1) & id4.mask
			for id4.allocated[pos] != NG {
				lh := h(id4.keys[pos]) & id4.mask

				if lh <= last && last < pos ||
					pos < lh && lh <= last ||
					last < pos && pos < lh {
					id4.keys[last] = id4.keys[pos]
					id4.allocated[last] = id4.allocated[pos]
					last = pos
				}
				pos = (pos + 1) & id4.mask
			}
			id4.keys[last] = 0
			id4.allocated[last] = NG

			return true
		}
		pos = (pos + 1) & id4.mask
	}
	return false
}

func (id4 *id4) id5(x, v int64) {
	nscale := id4.scale << 1
	nrscale := id4.rscale << 1
	nmask := nscale - 1
	nallocated := make([]int64, nscale)
	for i := 0; i < nscale; i++ {
		nallocated[i] = NG
	}
	nkeys := make([]int64, nscale)
	for i := id4.next(0); i < id4.scale; i = id4.next(i + 1) {
		y := id4.keys[i]
		pos := h(y) & nmask
		for nallocated[pos] != NG {
			pos = (pos + 1) & nmask
		}
		nkeys[pos] = y
		nallocated[pos] = id4.allocated[i]
	}
	{
		pos := h(x) & nmask
		for nallocated[pos] != NG {
			pos = (pos + 1) & nmask
		}
		nkeys[pos] = x
		nallocated[pos] = v
	}
	id4.allocated = nallocated
	id4.keys = nkeys
	id4.scale = nscale
	id4.rscale = nrscale
	id4.mask = nmask
}

func (id4 *id4) next(itr int) int {
	for itr < id4.scale && id4.allocated[itr] == NG {
		itr++
	}
	return itr
}

func h(x int64) int {
	x ^= x >> 33
	x *= 0xff51afd7ed558ccd
	x ^= x >> 33
	x *= 0xc4ceb9fe1a85ec53
	x ^= x >> 33
	return int(x)
}

type id5 struct {
	keys      []int64
	allocated []int64
	scale     int
	rscale    int
	mask      int
	size      int
}

func newID5() *id5 {
	id5 := &id5{}
	id5.scale = 1 << 2
	id5.rscale = 1 << 1
	id5.mask = id5.scale - 1
	id5.allocated = make([]int64, id5.scale)
	for i := 0; i < id5.scale; i++ {
		id5.allocated[i] = NG
	}
	id5.keys = make([]int64, id5.scale)
	return id5
}

func (id5 *id5) containsKey(x int64) bool {
	pos := h(x) & id5.mask
	for id5.allocated[pos] != NG {
		if x == id5.keys[pos] {
			return true
		}
		pos = (pos + 1) & id5.mask
	}
	return false
}

func (id5 *id5) get(x int64) int64 {
	pos := h(x) & id5.mask
	for id5.allocated[pos] != NG {
		if x == id5.keys[pos] {
			return id5.allocated[pos]
		}
		pos = (pos + 1) & id5.mask
	}
	return NG
}

func (id5 *id5) put(x, v int64) int64 {
	pos := h(x) & id5.mask
	for id5.allocated[pos] != NG {
		if x == id5.keys[pos] {
			oldval := id5.allocated[pos]
			id5.allocated[pos] = v
			return oldval
		}
		pos = (pos + 1) & id5.mask
	}
	if id5.size == id5.rscale {
		id5.id6(x, v)
	} else {
		id5.keys[pos] = x
		id5.allocated[pos] = v
	}
	id5.size++
	return 0
}

func (id5 *id5) inc(x, v int64) int64 {
	pos := h(x) & id5.mask
	for id5.allocated[pos] != NG {
		if x == id5.keys[pos] {
			id5.allocated[pos] += v
			return id5.allocated[pos]
		}
		pos = (pos + 1) & id5.mask
	}
	if id5.size == id5.rscale {
		id5.id6(x, v)
	} else {
		id5.keys[pos] = x
		id5.allocated[pos] = v
	}
	id5.size++
	return v
}

func (id5 *id5) remove(x int64) bool {
	pos := h(x) & id5.mask
	for id5.allocated[pos] != NG {
		if x == id5.keys[pos] {
			id5.size--

			last := pos
			pos = (pos + 1) & id5.mask
			for id5.allocated[pos] != NG {
				lh := h(id5.keys[pos]) & id5.mask

				if lh <= last && last < pos ||
					pos < lh && lh <= last ||
					last < pos && pos < lh {
					id5.keys[last] = id5.keys[pos]
					id5.allocated[last] = id5.allocated[pos]
					last = pos
				}
				pos = (pos + 1) & id5.mask
			}
			id5.keys[last] = 0
			id5.allocated[last] = NG

			return true
		}
		pos = (pos + 1) & id5.mask
	}
	return false
}

func (id5 *id5) id6(x, v int64) {
	nscale := id5.scale << 1
	nrscale := id5.rscale << 1
	nmask := nscale - 1
	nallocated := make([]int64, nscale)
	for i := 0; i < nscale; i++ {
		nallocated[i] = NG
	}
	nkeys := make([]int64, nscale)
	for i := id5.next(0); i < id5.scale; i = id5.next(i + 1) {
		y := id5.keys[i]
		pos := h(y) & nmask
		for nallocated[pos] != NG {
			pos = (pos + 1) & nmask
		}
		nkeys[pos] = y
		nallocated[pos] = id5.allocated[i]
	}
	{
		pos := h(x) & nmask
		for nallocated[pos] != NG {
			pos = (pos + 1) & nmask
		}
		nkeys[pos] = x
		nallocated[pos] = v
	}
	id5.allocated = nallocated
	id5.keys = nkeys
	id5.scale = nscale
	id5.rscale = nrscale
	id5.mask = nmask
}

func (id5 *id5) next(itr int) int {
	for itr < id5.scale && id5.allocated[itr] == NG {
		itr++
	}
	return itr
}

func h(x int64) int {
	x ^= x >> 33
	x *= 0xff51afd7ed558ccd
	x ^= x >> 33
	x *= 0xc4ceb9fe1a85ec53
	x ^= x >> 33
	return int(x)
}

const NG = 0

var picache = newID5()

func sqrt(n int64) int64 {
	i := int64(math.Max(0, math.Sqrt(float64(n))-2))
	for i*i <= n {
		i++
	}
	return i - 1
}

func id6(n int64) int64 {
	i := int64(math.Max(0, math.Cbrt(float64(n))-2))
	for i*i*i <= n {
		i++
	}
	return i - 1
}

func id0(n int) []int {
	if n <= 32 {
		primes := []int{2, 3, 5, 7, 11, 13, 17, 19, 23, 29, 31}
		for i := 0; i < len(primes); i++ {
			if n < primes[i] {
				return primes[:i]
			}
		}
		return primes
	}

	u := n + 32
	lu := math.Log(float64(u))
	ret := make([]int, int(u/lu+u/lu/lu*1.5))
	ret[0] = 2
	pos := 1

	isnp := make([]int, (n+1)/32/2+1)
	sup := (n + 1) / 32 / 2 + 1

	id3 := []int{3, 5, 7, 11, 13, 17, 19, 23, 29, 31}
	for _, tp := range id3 {
		ret[pos] = tp
		pos++
		ptn := make([]int, tp)
		for i := (tp - 3) / 2; i < tp<<5; i += tp {
			ptn[i>>5] |= 1 << (i & 31)
		}
		for j := 0; j < sup; j += tp {
			for i := 0; i < tp && i+j < sup; i++ {
				isnp[j+i] |= ptn[i]
			}
		}
	}

	magic := []int{0, 1, 23, 2, 29, 24, 19, 3, 30, 27, 25, 11, 20, 8, 4, 13, 31, 22, 28, 18, 26, 10, 7, 12, 21, 17, 9, 6, 16, 5, 15, 14}
	h := n / 2
	for i := 0; i < sup; i++ {
		for j := ^isnp[i]; j != 0; j &= j - 1 {
			pp := i<<5 | magic[(j&-j)*0x076be629>>27]
			p := 2*pp + 3
			if p > n {
				break
			}
			ret[pos] = p
			pos++
			if int64(p)*int64(p) > n {
				continue
			}
			for q := (p*p - 3) / 2; q <= h; q += p {
				isnp[q>>5] |= 1 << q
			}
		}
	}

	return ret[:pos]
}
